### DISCUSSION TITLE
Object condition handling

### DISCUSSION THREAD

**Discussion by owllex** [evennia/evennia#2454] [Evennia Questions] [answered]
I'd like to add support for temporary conditions on Objects (mostly Characters). Conditions in this case are coded effects that occur after a delay or on a timer (examples: a bleed effect that deals damage every few seconds, a temporary adjustment to a core statistic, or a disease that progresses over a long period of time). I suspect this is a fairly common kind of system.

The common component to all of these is time-based effects, either for when a condition wears off, for periodic ticks, or both.

*My question for the community is: what is a good way to architect a system like this?*

It seems to me that `delay()` ought to be able to cover the first case, and the global `TICKER_HANDLER` seems like a good fit for the second case.

I took a stab at a condition system by building a `ConditionHandler`, which was a light wrapper around an attribute attached to my `Character` typeclass, inspired by how Ainneve's `TraitHandler` works. I wrote a plain `Condition` class to contain the callback handling, and created subclasses of it to represent each kind of effect. `Conditions` are added to the `ConditionHandler` and then remove themselves when they expire. The idea was keep condition effects self-contained within their own object and not have to splatter their logic across the `Character` typeclass or other rules modules.

The problem I ran into with this approach is that the tickerhandler requires callbacks be registered on a typeclass or a global function. Ok, no problem, I can use the Character typeclass that owns the `ConditionHandler` to proxy a callback for me. Except, A) this is messy, and B) I keep running into errors when pickling those `Condition` objects and have to be very careful about what data they store. I could probably make this work with some more hacking, but it's starting to get a bit of a code smell to it and got me thinking about whether there's a better way.

Scripts seemed like an obvious choice for implementing Conditions, since they are pretty much made for the kind of callbacks I want to support. But I was worried the cost of repeatedly creating and destroying them might be too high for this. Some basic conditions really might just be modifying an attribute/tag when starting the condition, and then modifying it back on expiry, with some echoes in the middle. Creating a script that might only live for 10 seconds or so to do this feels heavy. Using a `delay()` seemed simpler, but there's something to be said for having all conditions built on the same system, as it keeps things cleaner.

How have others implemented this kind of system?

**Comment by Griatch** [comment, marked as answer]
@owllex The main question is if you need to use an active delay/timer at all. There is basically two possible cases here: 

- Possibility one: "_When the condition expires, I want to echo a message or do some _visible_ action that does not require player input or the engine to perform another action_".  
  Examples: 
  - "You are feeling hungry". 
  -  "The poison courses through your veins". 
  - "You are bleeding out"
- Possibility two: "_The effect of the expired condition will be noticeable only after active user input or engine action"_. 
  Examples: 
  - `action` -> "_When performing `<action>`, you feel a pang of hunger_", 
  - `hit <foe>` -> "_Your attack is weakened due to the poison coursing through your veins!_", 
  - `triage self` -> "_You investigate your wounds ... you have lost a lot of blood. You barely noticed._"
  - `<you get hit>` - gets shown the current HP, adjusted for poison along the way

For possibility one, you need active time keeping. A `delay` is usually the simplest way to do this. The [Command duration](https://www.evennia.com/docs/latest/Command-Duration.html) overview covers various ways to instill delays. 

For the second type (and, I argue, most cases) you _don't_ need any delays at all. Instead you do something like this (pseudo-code): 

```
# when effect starts 
start_time = time.time()
current_stat_value = 40   # whatever value we have right now, for example HP
max_allowed_stat_value = 100   # it can't go higher than this
effect_rate = 0.1  # per second (e.g. a heal)

# whenever checking the status (let's say it happens to be 10 seconds later)
current_time = time.time()
current_stat_value = min(max_allowed_stat_value, effect_rate * (current_time - start_time))
start_time = current_time
<display current_stat_value here>  # this is now 41. 
```

That is, you just track the rate of change of your effect and whenever you need to know it, you check how much time passed and figure out what the value must now be. It doesn't really need to be the user entering a command like above either - you could just have the combat engine get your current (modified) HP on-demand whenever it needs it for a calculation. 

In the `develop` branch of Evennia (what will become Evennia 1.0) there is a much extended implementation of the Ainneve Traits as a [formal Evennia contrib](https://github.com/evennia/evennia/blob/develop/evennia/contrib/traits.py). This implements on-demand timers like this as part of each Trait and while it's not quite ready for production yet it can be a good reference for seeing the principle.

**Reply by owllex** [reply]
I have a need for conditions in both categories, but your suggestion for continuous changes over time (fetched on property read) I think will work really well for some of my use cases, and wasn't something I had considered before, so thanks for the suggestion.

Heals over time (like active or passive HP/MP/etc recovery) seem like a great fit here since nothing in particular happens when those stats get topped off.

Damage over time though has the potential to kill a character, which could happen asynchronously and not just on property read. Right now my vitals module detects death in the HP setter, so either the dot effect needs to periodically check or write a new HP value, or it needs to try and predict eventual time of death (and then also respond to other effects or HP adjustments).

Maybe there's a better way to structure this, but I think no matter how I slice it, there's still a time-based component to any condition that has the potential to kill you.

**Comment by strikaco** [comment]
Heh, I just asked [a similar question](https://github.com/evennia/evennia/discussions/2406) not that long ago.

Your best bet is going to be scripts-- timers are a native function and no more worries about pickling.

I would also recommend implementing a semaphore (the traits handler does this with the `mod` concept)--

* On the target object, have a _numeric_ counter attribute for the various conditions possible-- `is_poisoned=0`, `is_drunk=0`, etc.
* Every time you apply a condition, add 1 to the appropriate `is_x` attribute.
* When condition ends, subtract 1 from the appropriate attribute.

This way, multiple sources can apply `poison` independent of each other-- and the first one to expire won't cancel out the rest. The scripts themselves are free to apply the condition in any way they want; two `poison` scripts can be in effect simultaneously and cause damage at different rates.

And you can easily tell whether an object is affected by _any form of the given condition_ by checking the truthiness of the appropriate `is_x` attribute. This should pair seamlessly with the default lock functions (`traverse:not objattr(is_immobile)`).

Griatch's suggestion will prevent unnecessary hammering of the db (otherwise any sort of HoT/DoT condition is going to do a db write every _n_ seconds...for every single object currently affected by such a condition).

**Reply by owllex** [reply]
Did you end up creating a separate typeclass (inheriting from Script) for each kind of condition you support? Or did you create a single parent `Condition` typeclass, and then bolt on condition-specific logic into that (either from separate classes, or though some omni-condition handler)?

**Reply by strikaco** [reply]
I tried that initially and hated it-- too many typeclasses to keep track of, and made for a very rigid architecture.

Currently one Condition script with a bunch of attributes and methods directing enactment is working well for me. This also allows construction of complex conditions through selection of effects alone.

The code redundancy can be reduced any number of ways, but the idea is:

```
class Condition(Script):

    # semaphores:
    # bool(obj.db.is_drunk) will always show whether the given obj is currently drunk by any means
    def at_start(self):
        if self.db.drunk:
            self.obj.db.is_drunk += 1
        # etc

    def at_stop(self):
        if self.db.drunk:
            self.obj.db.is_drunk -= 1
        # etc

    def at_repeat(self):
        if self.db.poison:
            self.at_poison_tick()
        if self.db.burning:
            self.at_burning_tick()
        if self.db.drunk:
            self.at_drunk_tick()
        # etc

    def at_burning_tick(self):
        self.obj.db.hp -= 5
        self.obj.msg("Your untreated gonorrhea is causing you internal tissue damage.")

    def at_drunk_tick(self):
        import random
        dest = random.choice(self.obj.location.exits)
        self.obj.location.msg_contents(f"{self.obj.get_display_name(None)} trips over their own feet and stumbles toward {dest}.")
        self.obj.move_to(dest)

```

**Reply by owllex** [reply]
Thanks, this is helpful.
